perllib/FixMyStreet/DB/ResultSet/ResponsePriority.pm

```diff
@@ -3,20 +3,39 @@ use base 'DBIx::Class::ResultSet';
 
 use strict;
 use warnings;
+use JSON::MaybeXS;
 
 sub for_bodies {
     my ($rs, $bodies, $category) = @_;
     my $attrs = {
         'me.body_id' => $bodies,
     };
+    my $filters = {
+      order_by => 'name',
+      join => { 'contact_response_priorities' => 'contact' },
+      distinct => 1,
+    };
     if ($category) {
         $attrs->{'contact.category'} = [ $category, undef ];
     }
-    $rs->search($attrs, {
-        order_by => 'name',
-        join => { 'contact_response_priorities' => 'contact' },
-        distinct => 1,
-    });
+    $rs->search($attrs, $filters);
+}
+
+sub by_categories {
+    my ($rs, $area_id, @contacts) = @_;
+    my %body_ids = map { $_->body_id => 1 } FixMyStreet::DB->resultset('BodyArea')->search({ area_id => $area_id });
+    my @body_ids = keys %body_ids;
+    my %priorities = ();
+    my @results = $rs->for_bodies(\@body_ids, undef);
+    @contacts = grep { $body_ids{$_->body_id} } @contacts;
+
+    foreach my $contact (@contacts) {
+      my @ts = grep { $_->contact_response_priorities->find({contact_id => $contact->get_column('id')}) } @results;
```

Comment:
You have not made the change here that we discussed at https://mysociety.slack.com/archives/D41T2K7T5/p1494601362486772 - this line should be the same as the defect types code.
In fact, these two functions are practically identical, can you factor them out to one function? The only difference looks to be the many-many lookup table used, and that defect types include `extra` in their output and priorities do not. You could use a Role to store the shared function (ie. put it under FixMyStreet/Roles and use `with` in the two ResultSet files). The same presumably applies to `for_bodies`.